### ISSUE TITLE
filter query parameters of old AiiDA REST API incompatible with OpenAPI/fastapi standard

### ISSUE THREAD

**Issue by ltalirz** [aiidateam/aiida-restapi#16]
It seems that, in general, the [exact structure of the query string is not standardized](https://en.wikipedia.org/wiki/Query_string#Structure).
For example, [RFC3986 only mentions that](https://datatracker.ietf.org/doc/html/rfc3986#section-3.4)
> query components are often used to carry identifying information in the form of "key=value" pairs

Similar the latest [RFC8820](https://datatracker.ietf.org/doc/html/rfc8820#section-2.4).

However, there is a [W3C recommendation for encoding form data (`application/x-www-form-urlencoded`)](https://www.w3.org/TR/REC-html40/interact/forms.html#form-content-type):
> The name is separated from the value by `=` and name/value pairs are separated from each other by `&`

While the OpenAPI standard is not explicit on the name-value separator, all of the [examples in the spec](https://swagger.io/docs/specification/serialization/#query) use `=`.

As pointed out by @chrisjsewell, `fastapi` relies on `starlette` which [internally](https://github.com/encode/starlette/blob/b6f3578bb2cf6c60e3efe110143409b47f368d36/starlette/datastructures.py#L401) relies on `urllib`'s `parse_qsl` function for parsing the query parameters. `parse_qsl` (part of the python standard library) also [hardcodes the `=` separator](https://github.com/python/cpython/blob/320eaa7f42b413cd5e5436ec92d4dc5ba150395f/Lib/urllib/parse.py#L751).

This clashes with the [filter operators](https://aiida.readthedocs.io/projects/aiida-core/en/latest/reference/rest_api.html#filter-operators) from the aiida-core rest api, which allow e.g. for queries like
```
http://localhost:5000/api/v4/nodes?mtime>=2019-04-23
```

**Comment by ltalirz** [comment]
One way around it could be to introduce a `filter` parameter, e.g.
```
http://localhost:5000/api/v4/nodes?filter="mtime>=2019-04-23"
```

It seems to me that this should work in principle, as the [openapi standard states](https://swagger.io/docs/specification/serialization/#query)

> Additionally, the allowReserved keyword specifies whether the reserved characters :/?#[]@!$&'()*+,;= in parameter values are allowed to be sent as they are, or should be percent-encoded. By default, allowReserved is false, and reserved characters are percent-encoded. 

However, looking at the string that is fed into `parse_qsl`, I notice that e.g. `/` is encoded while `&` is not:
```
# GET http://127.0.0.1:8000/users?id="123&23"
id=%22123&23%22
INFO:     127.0.0.1:56308 - "GET /users?id=%22123&23%22 HTTP/1.1" 200 OK
```
This means applying `parse_qsl` on that string leads to an unwanted split at the `&` symbol.

P.S. Hmm... the funny thing is that even copying the URL from the URL field of the web browser into this text box already encodes `"` but not `&`: http://127.0.0.1:8000/users?id=%22123&23%22
="1eee&4"

If that's the way things are supposed to work, then that solution doesn't work (of course one could just specify `filter=` multiple times).

**Comment by chrisjsewell** [comment]
Oh yeh indeed, its a bit of a pain to use as a "client", you can see here where I have to do "partial" encoding to get it to work:

https://github.com/chrisjsewell/react-aiida-app/blob/6722e8e4dea489315064fe06f94c8cf3015ebde0/src/clients/aiidaClient.ts#L88-L91

So yeh, in terms of not breaking too much from the existing API, I can see why we want to implement this filter parameter, but... from a top-level / client perspective, I feel I would just generally avoid it, and use instead the "more structured" querybuilder/graphql endpoint

**Comment by ltalirz** [comment]
I just discovered that fastapi has built-in support for parsing the starlette request object manually: https://fastapi.tiangolo.com/advanced/using-request-directly/

If I understand correctly, one can even still have automatic parsing & docs for all of the fields that follow the `key=value` logic.

**Comment by chrisjsewell** [comment]
> It would also mean that if you get data from the Request object directly (for example, read the body) it won't be validated, converted or documented (with OpenAPI, for the automatic API user interface) by FastAPI.

Interesting cheers, although... it sounds like that removes a lot of the features of fastapi

**Comment by ltalirz** [comment]
If I understand correctly, then this is a limitation of the OpenAPI standard: query parameters can only *have* a value, there is no concept of different relations  (smaller than, ...) between query parameters and their value other than equality; i.e. there is nothing `fastapi` can do about it.

This would mean that the filter parameters would be missing e.g. from the interactive client built into the docs; on the other hand it would potentially allow us to be fully backward compatible, which is worth something.

**Comment by chrisjsewell** [comment]
Well it would not just be the filter parameters, it would be every parameter, wouldn't it?
i.e once you have take the onus to parse the query string then fastapi will have no idea what any of it is

**Comment by chrisjsewell** [comment]
On a high-level note, it would be good to have an idea of who is already using the REST API, and so who this would affect (just Mat Cloud or lots of others). It does clearly state when you start it presently that it should not be used for production use, but then it also has been around for many years.

**Comment by ltalirz** [comment]
> Well it would not just be the filter parameters, it would be every parameter, wouldn't it?

I don't think so - as mentioned [in the page I linked](https://fastapi.tiangolo.com/advanced/using-request-directly/), you can simply declare any of the parameters the use the `=` and then do the manual parsing of the Request for the rest.

> On a high-level note, it would be good to have an idea of who is already using the REST API, and so who this would affect (just Mat Cloud or lots of others).

That is a good point. So far I know of
 * the AiiDA explorer https://github.com/materialscloud-org/aiida-explorer
 * aiida-post https://github.com/d-tomerini/aiida-post
 * your jlab-aiidatree https://github.com/chrisjsewell/jlab_aiidatree/
 * this archived "materials simulation environment" https://github.com/anoopkcn/mse/blob/78115d153d12c3865f6511046cbd9dd74525deb6/src/components/Computers.jsx
 * some AiiDA tutorials

There may be a little more, and we should ask around in a few days once we have a clearer idea of what we could realistically do. I do suspect that the migration of the aiida-explorer would be the largest task.

> It does clearly state when you start it presently that it should not be used for production use, but then it also has been around for many years.

Well, that is just for the server that ships with `flask`.
On Materials Cloud we run it with apache via `mod_wsgi`.

**Comment by chrisjsewell** [comment]
> one would need to check that the parsing works for the "mixed" query strings, but I'm optimistic since it should split on & which should work nevertheless

well I will leave you to figure that out for now, because yeh, from a selfish viewpoint, if I can get the graphql working nicely, I will just look to switch to that for all client-side querying 😝

**Comment by edan-bainglass** [comment]
Two comments:
1. Though GraphQL is indeed powerful and standard for queries, there is arguably still value to providing simple filtering at `get_many`-type endpoints. Currently, we parse the dict filters from the URL - not great, and indeed a pain for (human) clients.
2. For standardizing the filters (and sorting, pagination, projections), we can look to [JSON:API](https://jsonapi.org/format/#query-parameters-families) (see #18, superseded by #101). There seems to be some guidelines also [here](https://www.drupal.org/docs/core-modules-and-themes/core-modules/jsonapi-module/filtering), or at least one implementation.

**Comment by edan-bainglass** [comment]
Superseded by https://github.com/aiidateam/aiida-restapi/issues/163
